### ISSUE TITLE
Detach git clone from the `GitRepo` initializer and other behavior changes.

### ISSUE THREAD

**Issue by minuscorp** [fastlane/ci#371]
I really think that the API build over the `GitRepo` class is kinda confusing, at least for myself. Between some of the issues one of the most obvious is the clone of the repo being done on the `initializer` of the repo. I really think that this shouldn't be done this way, as it provides a confusing behavior not expected by the sense of the method itself.

I have some ideas in mind in following PRs, but before starting to work I'd like to get some feedback:

- First of all, detach the `clone` from the `initializer`. This should be manually triggered by the developer. Also, the `GitRepo` should store its internal state on load. (i.e., a `GitRepo` on initialization time should know if the repo is already cloned, which `branch` is checked out, and what commit `sha` it is).

- Second of all, a `GitRepo` can have several states on it self through the `repo_config` property. We have a `root_path`, but if a `Project` uses a single `repo_config` and uses several triggers (manually, commit, nightly) the `GitRepo` should have sub-clones that depends on its internal state.
     1. I want my `GitRepo` being `setup` on `branch` `master`, the `clone_path` then should be `root_path/master/latest_sha_of_master/`
     1. I want my `GitRepo` being `setup` on `sha` `abcd`, the `clone_path` then should be `root_path/branch_of_the_sha/abcd/`
     1. I do not provide any `setup` information, the `clone_path` then should be `root_path/master/latest_sha_of_master/`.

- The behavior of the current `checkout_branch` and `checkout_commit` behaves the same way provided above. So from `root_path`, every variant of `branch` and `sha` are built.

- `clear_directory` would behave as it does now, deleting all the `root_path` tree. but we can extend it to `clear_directory(sha: nil, branch: nil)` which will delete just certain paths.
    1. `clear_directory(sha: "abcd")` will delete `root_path/branch_of_the_sha/abcd`.
    1. `clear_directory(branch: "master")` will delete `root_path/master/**`.

I'm really into this little refactor (which will break almost anything of the current setup) and provides a more robust and reliant architecture).

### Benefits
1. Won't need to have conflicts when different workers might be working with the same `repo_config` but different `sha`, `branch` or general state of the repo itself.
1. All the explanation given above.

### Cons
1. Cloning won't longer be on the `initializer`.
1. New clone api: `clone(branch: nil, sha: nil, shallow: true)`.

cc @taquitos @KrauseFx

**Comment by minuscorp** [comment]
Discussion on this? We can manage to split this into small pieces and don't end up with a massive PR. Let me know if a call is required if something is not clearly explained here.

**Comment by minuscorp** [comment]
I ping again @taquitos and @KrauseFx , should this be done for beta? Or just close this
